File: backend/routes/recruiter.py
from fastapi import APIRouter, HTTPException, Depends, Query
from pydantic import BaseModel
from typing import Optional, List
import json
from database import get_db
from auth import get_current_user

router = APIRouter()


class StatusUpdateReq(BaseModel):
    status: str
    notes: Optional[str] = None


def clean_stale_sessions(conn):
    """Automatically convert stale abandoned in_progress or paused sessions (older than 1 hour) into completed state."""
    conn.execute("""
        UPDATE interview_session
        SET status = 'completed',
            completed_at = COALESCE(started_at, created_at)
        WHERE status IN ('in_progress', 'paused')
          AND (
            (started_at IS NOT NULL AND strftime('%s', 'now') - strftime('%s', started_at) > 3600)
            OR (started_at IS NULL AND strftime('%s', 'now') - strftime('%s', created_at) > 3600)
          )
    """)
    conn.commit()


@router.get("/summary")
def get_recruiter_summary(user: dict = Depends(get_current_user)):
    user_role = user.get("role", "candidate")
    if user_role not in ("recruiter", "admin"):
        raise HTTPException(status_code=403, detail="Access denied. Recruiter role required.")

    conn = get_db()
    clean_stale_sessions(conn)
    recruiter_id = user["id"]

    # 1. Total Candidates
    total_candidates = conn.execute("SELECT COUNT(*) FROM users WHERE role = 'candidate'").fetchone()[0]

    # 2. Shortlisted Count for this recruiter
    shortlisted_count = conn.execute(
        "SELECT COUNT(*) FROM recruiter_candidate_status WHERE recruiter_id = ? AND status = 'shortlisted'",
        (recruiter_id,)
    ).fetchone()[0]

    # 3. Active Live Sessions
    active_live_sessions = conn.execute(
        "SELECT COUNT(*) FROM interview_session WHERE status IN ('in_progress', 'paused')"
    ).fetchone()[0]

    # 4. Platform Average Candidate Score
    avg_row = conn.execute(
        "SELECT AVG(overall_score) FROM interview_session WHERE overall_score IS NOT NULL AND status = 'completed'"
    ).fetchone()
    avg_score = round(avg_row[0], 1) if avg_row and avg_row[0] is not None else 0.0

    # 5. Hiring Funnel Breakdown
    assessed_count = conn.execute(
        "SELECT COUNT(DISTINCT candidate_id) FROM interview_session WHERE status = 'completed'"
    ).fetchone()[0]

    under_review_count = conn.execute(
        "SELECT COUNT(*) FROM recruiter_candidate_status WHERE recruiter_id = ? AND status = 'under_review'",
        (recruiter_id,)
    ).fetchone()[0]

    rejected_count = conn.execute(
        "SELECT COUNT(*) FROM recruiter_candidate_status WHERE recruiter_id = ? AND status = 'rejected'",
        (recruiter_id,)
    ).fetchone()[0]

    hiring_funnel = {
        "applied": total_candidates,
        "assessed": assessed_count,
        "shortlisted": shortlisted_count,
        "under_review": under_review_count,
        "rejected": rejected_count
    }

    # 6. Top Candidates Spotlight
    top_candidates_rows = conn.execute("""
        SELECT 
            u.id, u.name, u.email,
            AVG(s.overall_score) as avg_overall,
            AVG(s.technical_score) as avg_tech,
            COUNT(s.id) as total_sessions,
            MAX(s.domain) as last_domain,
            COALESCE(rcs.status, 'new') as shortlist_status
        FROM users u
        JOIN interview_session s ON u.id = COALESCE(s.candidate_id, s.user_id)
        LEFT JOIN recruiter_candidate_status rcs ON rcs.candidate_id = u.id AND rcs.recruiter_id = ?
        WHERE u.role = 'candidate' AND s.overall_score IS NOT NULL
        GROUP BY u.id
        ORDER BY avg_overall DESC
        LIMIT 5
    """, (recruiter_id,)).fetchall()

    top_candidates = []
    for r in top_candidates_rows:
        top_candidates.append({
            "id": r["id"],
            "name": r["name"],
            "email": r["email"],
            "overall_score": round(r["avg_overall"], 1) if r["avg_overall"] else 0,
            "technical_score": round(r["avg_tech"], 1) if r["avg_tech"] else 0,
            "total_sessions": r["total_sessions"],
            "domain": r["last_domain"] or "Software Engineering",
            "status": r["shortlist_status"]
        })

    conn.close()

    return {
        "total_candidates": total_candidates,
        "shortlisted_count": shortlisted_count,
        "active_live_sessions": active_live_sessions,
        "avg_candidate_score": avg_score,
        "hiring_funnel": hiring_funnel,
        "top_candidates": top_candidates
    }


@router.get("/candidates")
def get_recruiter_candidates(
    search: Optional[str] = None,
    status_filter: Optional[str] = Query("all"),
    user: dict = Depends(get_current_user)
):
    user_role = user.get("role", "candidate")
    if user_role not in ("recruiter", "admin"):
        raise HTTPException(status_code=403, detail="Access denied. Recruiter role required.")

    conn = get_db()
    recruiter_id = user["id"]

    query = """
        SELECT 
            u.id, u.name, u.email, u.created_at,
            COALESCE(rcs.status, 'new') as shortlist_status,
            rcs.notes,
            COUNT(DISTINCT s.id) as sessions_count,
            COUNT(DISTINCT a.id) as assessments_count,
            AVG(s.overall_score) as avg_overall,
            AVG(s.technical_score) as avg_technical,
            AVG(s.communication_score) as avg_communication,
            AVG(s.confidence_score) as avg_confidence,
            AVG(a.score_percentage) as avg_assessment,
            MAX(s.domain) as domain,
            MAX(s.created_at) as last_interview_date
        FROM users u
        LEFT JOIN recruiter_candidate_status rcs ON rcs.candidate_id = u.id AND rcs.recruiter_id = ?
        LEFT JOIN interview_session s ON (u.id = s.candidate_id OR u.id = s.user_id) AND s.status = 'completed'
        LEFT JOIN assessment a ON u.id = a.user_id AND a.status = 'completed'
        WHERE u.role = 'candidate'
    """
    params = [recruiter_id]

    if search:
        query += " AND (u.name LIKE ? OR u.email LIKE ? OR s.domain LIKE ?)"
        term = f"%{search.strip()}%"
        params.extend([term, term, term])

    if status_filter and status_filter != "all":
        if status_filter == "new":
            query += " AND rcs.status IS NULL"
        else:
            query += " AND rcs.status = ?"
            params.append(status_filter)

    query += " GROUP BY u.id ORDER BY avg_overall DESC, u.created_at DESC"

    rows = conn.execute(query, params).fetchall()
    candidates = []

    for r in rows:
        overall = round(r["avg_overall"], 1) if r["avg_overall"] is not None else (round(r["avg_assessment"], 1) if r["avg_assessment"] is not None else 0.0)
        tech = round(r["avg_technical"], 1) if r["avg_technical"] is not None else 0.0
        comm = round(r["avg_communication"], 1) if r["avg_communication"] is not None else 0.0
        conf = round(r["avg_confidence"], 1) if r["avg_confidence"] is not None else 0.0
        assess = round(r["avg_assessment"], 1) if r["avg_assessment"] is not None else 0.0

        if overall >= 85:
            rating = "Exceptional"
            trend = "improving"
        elif overall >= 70:
            rating = "Strong"
            trend = "stable"
        elif overall >= 50:
            rating = "Satisfactory"
            trend = "stable"
        elif overall > 0:
            rating = "Needs Work"
            trend = "needs_work"
        else:
            rating = "Not Assessed"
            trend = "stable"

        candidates.append({
            "id": r["id"],
            "name": r["name"],
            "email": r["email"],
            "created_at": r["created_at"],
            "status": r["shortlist_status"],
            "notes": r["notes"] or "",
            "sessions_count": r["sessions_count"],
            "assessments_count": r["assessments_count"],
            "overall_score": overall,
            "technical_score": tech,
            "communication_score": comm,
            "confidence_score": conf,
            "assessment_score": assess,
            "domain": r["domain"] or "General Practice",
            "rating": rating,
            "trend": trend,
            "last_active": r["last_interview_date"] or r["created_at"]
        })

    conn.close()
    return {"candidates": candidates}


@router.put("/candidates/{candidate_id}/status")
def update_candidate_status(
    candidate_id: int,
    req: StatusUpdateReq,
    user: dict = Depends(get_current_user)
):
    user_role = user.get("role", "candidate")
    if user_role not in ("recruiter", "admin"):
        raise HTTPException(status_code=403, detail="Access denied. Recruiter role required.")

    if req.status not in ("shortlisted", "under_review", "rejected"):
        raise HTTPException(status_code=400, detail="Invalid status value.")

    conn = get_db()
    recruiter_id = user["id"]

    cand = conn.execute("SELECT id, name FROM users WHERE id = ? AND role = 'candidate'", (candidate_id,)).fetchone()
    if not cand:
        conn.close()
        raise HTTPException(status_code=404, detail="Candidate not found.")

    conn.execute("""
        INSERT INTO recruiter_candidate_status (recruiter_id, candidate_id, status, notes, updated_at)
        VALUES (?, ?, ?, ?, CURRENT_TIMESTAMP)
        ON CONFLICT(recruiter_id, candidate_id) DO UPDATE SET
            status = excluded.status,
            notes = COALESCE(excluded.notes, recruiter_candidate_status.notes),
            updated_at = CURRENT_TIMESTAMP
    """, (recruiter_id, candidate_id, req.status, req.notes))
    conn.commit()
    conn.close()

    return {"message": f"Candidate status updated to {req.status}", "candidate_id": candidate_id, "status": req.status}


class TemplateCreateReq(BaseModel):
    title: str
    interview_type: str
    domain: Optional[str] = None
    difficulty: Optional[str] = "medium"
    duration_minutes: Optional[int] = 15
    num_questions: Optional[int] = 5
    topics: Optional[List[str]] = []
    description: Optional[str] = None


@router.get("/sessions")
def get_recruiter_sessions(
    status_filter: Optional[str] = Query("all"),
    user: dict = Depends(get_current_user)
):
    user_role = user.get("role", "candidate")
    if user_role not in ("recruiter", "admin"):
        raise HTTPException(status_code=403, detail="Access denied. Recruiter role required.")

    conn = get_db()
    clean_stale_sessions(conn)
    query = """
        SELECT 
            s.id as session_id,
            s.interview_type,
            s.domain,
            s.difficulty,
            s.status,
            s.duration,
            s.elapsed_seconds,
            s.current_question_index,
            s.overall_score,
            s.technical_score,
            s.communication_score,
            s.confidence_score,
            s.performance_rating,
            s.created_at,
            s.started_at,
            s.completed_at,
            u.id as candidate_id,
            u.name as candidate_name,
            u.email as candidate_email,
            r.id as recording_id,
            r.mime_type as recording_mime_type,
            r.file_size_bytes as recording_size
        FROM interview_session s
        JOIN users u ON u.id = COALESCE(s.candidate_id, s.user_id)
        LEFT JOIN interview_recording r ON r.session_id = s.id
        WHERE u.role = 'candidate'
    """
    params = []

    if status_filter and status_filter != "all":
        if status_filter == "live":
            query += " AND s.status IN ('in_progress', 'paused')"
        else:
            query += " AND s.status = ?"
            params.append(status_filter)

    query += " ORDER BY s.created_at DESC"
    rows = conn.execute(query, params).fetchall()

    sessions = []
    for r in rows:
        q_count = conn.execute(
            "SELECT COUNT(*) FROM interview_question WHERE interview_id = ?",
            (r["session_id"],)
        ).fetchone()[0]

        q_curr = conn.execute(
            "SELECT question_text FROM interview_question WHERE interview_id = ? AND sequence_no = ?",
            (r["session_id"], (r["current_question_index"] or 0) + 1)
        ).fetchone()

        sessions.append({
            "session_id": r["session_id"],
            "interview_type": r["interview_type"],
            "domain": r["domain"] or "General Domain",
            "difficulty": r["difficulty"] or "medium",
            "status": r["status"],
            "duration": r["duration"] or 15,
            "elapsed_seconds": r["elapsed_seconds"] or 0,
            "current_question_index": r["current_question_index"] or 0,
            "total_questions": q_count or 5,
            "current_question": q_curr["question_text"] if q_curr else "N/A",
            "overall_score": round(r["overall_score"], 1) if r["overall_score"] is not None else None,
            "technical_score": round(r["technical_score"], 1) if r["technical_score"] is not None else None,
            "communication_score": round(r["communication_score"], 1) if r["communication_score"] is not None else None,
            "confidence_score": round(r["confidence_score"], 1) if r["confidence_score"] is not None else None,
            "performance_rating": r["performance_rating"],
            "created_at": r["created_at"],
            "started_at": r["started_at"],
            "completed_at": r["completed_at"],
            "candidate_id": r["candidate_id"],
            "candidate_name": r["candidate_name"],
            "candidate_email": r["candidate_email"],
            "recording_id": r["recording_id"],
            "recording_mime_type": r["recording_mime_type"],
            "recording_size": r["recording_size"]
        })

    conn.close()
    return {"sessions": sessions}


@router.get("/compare")
def get_recruiter_compare(
    candidate_ids: str = Query(..., description="Comma separated candidate user IDs"),
    user: dict = Depends(get_current_user)
):
    user_role = user.get("role", "candidate")
    if user_role not in ("recruiter", "admin"):
        raise HTTPException(status_code=403, detail="Access denied. Recruiter role required.")

    try:
        c_ids = [int(x.strip()) for x in candidate_ids.split(",") if x.strip().isdigit()]
    except Exception:
        raise HTTPException(status_code=400, detail="Invalid candidate IDs.")

    if not c_ids:
        return {"comparison": []}

    conn = get_db()
    recruiter_id = user["id"]
    placeholders = ",".join(["?"] * len(c_ids))

    query = f"""
        SELECT 
            u.id, u.name, u.email,
            COALESCE(rcs.status, 'new') as shortlist_status,
            COUNT(DISTINCT s.id) as sessions_count,
            AVG(s.overall_score) as avg_overall,
            AVG(s.technical_score) as avg_technical,
            AVG(s.communication_score) as avg_communication,
            AVG(s.confidence_score) as avg_confidence,
            AVG(s.professionalism_score) as avg_professionalism,
            AVG(a.score_percentage) as avg_assessment,
            MAX(s.domain) as domain
        FROM users u
        LEFT JOIN recruiter_candidate_status rcs ON rcs.candidate_id = u.id AND rcs.recruiter_id = ?
        LEFT JOIN interview_session s ON (u.id = s.candidate_id OR u.id = s.user_id) AND s.status = 'completed'
        LEFT JOIN assessment a ON u.id = a.user_id AND a.status = 'completed'
        WHERE u.id IN ({placeholders}) AND u.role = 'candidate'
        GROUP BY u.id
    """
    params = [recruiter_id] + c_ids
    rows = conn.execute(query, params).fetchall()

    comparison = []
    for r in rows:
        tech = round(r["avg_technical"], 1) if r["avg_technical"] is not None else 0.0
        comm = round(r["avg_communication"], 1) if r["avg_communication"] is not None else 0.0
        conf = round(r["avg_confidence"], 1) if r["avg_confidence"] is not None else 0.0
        prof = round(r["avg_professionalism"], 1) if r["avg_professionalism"] is not None else 0.0
        assess = round(r["avg_assessment"], 1) if r["avg_assessment"] is not None else 0.0
        overall = round(r["avg_overall"], 1) if r["avg_overall"] is not None else (assess or 0.0)

        # AI Recommendation Synthesis
        if overall >= 85:
            fit = "Strong Hiring Candidate — High Technical & Communication Aptitude"
            badge = "Top Pick"
        elif overall >= 70:
            fit = "Solid Candidate — Recommending Final Round Technical Review"
            badge = "Good Match"
        elif overall >= 50:
            fit = "Moderate Fit — Additional Skills Assessment Advised"
            badge = "Moderate"
        else:
            fit = "Requires Skill Enhancement & Further Training"
            badge = "Needs Work"

        comparison.append({
            "id": r["id"],
            "name": r["name"],
            "email": r["email"],
            "domain": r["domain"] or "General Engineering",
            "status": r["shortlist_status"],
            "overall_score": overall,
            "technical_score": tech,
            "communication_score": comm,
            "confidence_score": conf,
            "professionalism_score": prof,
            "assessment_score": assess,
            "sessions_count": r["sessions_count"],
            "ai_recommendation": fit,
            "badge": badge
        })

    conn.close()
    return {"comparison": comparison}


@router.get("/sessions/live")
def get_recruiter_live_sessions(user: dict = Depends(get_current_user)):
    return get_recruiter_sessions(status_filter="live", user=user)


# Templates API
@router.get("/templates")
def get_recruiter_templates(user: dict = Depends(get_current_user)):
    user_role = user.get("role", "candidate")
    if user_role not in ("recruiter", "admin"):
        raise HTTPException(status_code=403, detail="Access denied. Recruiter role required.")

    conn = get_db()
    recruiter_id = user["id"]

    rows = conn.execute("""
        SELECT * FROM interview_template 
        WHERE recruiter_id IS NULL OR recruiter_id = ?
        ORDER BY id DESC
    """, (recruiter_id,)).fetchall()

    templates = []
    for r in rows:
        topics = []
        if r["topics_json"]:
            try:
                topics = json.loads(r["topics_json"])
            except Exception:
                topics = []

        templates.append({
            "id": r["id"],
            "recruiter_id": r["recruiter_id"],
            "title": r["title"],
            "interview_type": r["interview_type"],
            "domain": r["domain"] or "Software Engineering",
            "difficulty": r["difficulty"] or "medium",
            "duration_minutes": r["duration_minutes"] or 15,
            "num_questions": r["num_questions"] or 5,
            "topics": topics,
            "description": r["description"] or "",
            "is_system": r["recruiter_id"] is None,
            "created_at": r["created_at"]
        })

    conn.close()
    return {"templates": templates}


@router.post("/templates")
def create_recruiter_template(req: TemplateCreateReq, user: dict = Depends(get_current_user)):
    user_role = user.get("role", "candidate")
    if user_role not in ("recruiter", "admin"):
        raise HTTPException(status_code=403, detail="Access denied. Recruiter role required.")

    if not req.title.strip() or not req.interview_type.strip():
        raise HTTPException(status_code=400, detail="Title and interview type are required.")

    conn = get_db()
    recruiter_id = user["id"]
    topics_json = json.dumps(req.topics or [])

    cur = conn.execute("""
        INSERT INTO interview_template (recruiter_id, title, interview_type, domain, difficulty, duration_minutes, num_questions, topics_json, description)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
    """, (
        recruiter_id,
        req.title.strip(),
        req.interview_type.strip(),
        req.domain.strip() if req.domain else "Software Engineering",
        req.difficulty or "medium",
        req.duration_minutes or 15,
        req.num_questions or 5,
        topics_json,
        req.description.strip() if req.description else ""
    ))
    conn.commit()
    template_id = cur.lastrowid
    row = conn.execute("SELECT * FROM interview_template WHERE id = ?", (template_id,)).fetchone()
    conn.close()

    return {
        "message": "Interview template created successfully.",
        "template": {
            "id": row["id"],
            "recruiter_id": row["recruiter_id"],
            "title": row["title"],
            "interview_type": row["interview_type"],
            "domain": row["domain"],
            "difficulty": row["difficulty"],
            "duration_minutes": row["duration_minutes"],
            "num_questions": row["num_questions"],
            "topics": req.topics or [],
            "description": row["description"],
            "is_system": False,
            "created_at": row["created_at"]
        }
    }


@router.delete("/templates/{template_id}")
def delete_recruiter_template(template_id: int, user: dict = Depends(get_current_user)):
    user_role = user.get("role", "candidate")
    if user_role not in ("recruiter", "admin"):
        raise HTTPException(status_code=403, detail="Access denied. Recruiter role required.")

    conn = get_db()
    recruiter_id = user["id"]

    tpl = conn.execute("SELECT id, recruiter_id FROM interview_template WHERE id = ?", (template_id,)).fetchone()
    if not tpl:
        conn.close()
        raise HTTPException(status_code=404, detail="Template not found.")

    if tpl["recruiter_id"] is None and user_role != "admin":
        conn.close()
        raise HTTPException(status_code=403, detail="Cannot delete default system template.")

    conn.execute("DELETE FROM interview_template WHERE id = ?", (template_id,))
    conn.commit()
    conn.close()

    return {"message": "Template deleted successfully.", "template_id": template_id}

File: backend/routes/recruiter.js
// backend/routes/recruiter.js
const express = require('express');
const router = express.Router();
const pool = require('../db');
const { auth, checkRole } = require('../middleware/auth');

// =============================================
// GET ALL CANDIDATES WITH THEIR INTERVIEW DATA
// =============================================
router.get('/candidates', auth, checkRole(['RECRUITER', 'ADMIN']), async (req, res) => {
  try {
    const { search, status, sortBy = 'avg_score', sortOrder = 'DESC' } = req.query;

    // Base query: Get all users with USER role and their interview data
    let query = `
      SELECT 
        u.id,
        u.name,
        u.email,
        u.created_at as joined_date,
        COUNT(DISTINCT i.id) as total_interviews,
        COUNT(DISTINCT CASE WHEN i.status = 'completed' THEN i.id END) as completed_interviews,
        AVG(CASE WHEN i.status = 'completed' THEN i.score END) as avg_score,
        MAX(CASE WHEN i.status = 'completed' THEN i.score END) as highest_score,
        MIN(CASE WHEN i.status = 'completed' THEN i.score END) as lowest_score,
        MAX(CASE WHEN i.status = 'completed' THEN i.created_at END) as last_interview_date,
        COUNT(DISTINCT CASE WHEN i.status = 'in_progress' THEN i.id END) as active_sessions,
        (
          SELECT JSON_AGG(
            JSON_BUILD_OBJECT(
              'id', i2.id,
              'type', i2.interview_type,
              'domain', i2.domain,
              'difficulty', i2.difficulty,
              'score', i2.score,
              'status', i2.status,
              'created_at', i2.created_at,
              'feedback', i2.feedback
            )
            ORDER BY i2.created_at DESC
          )
          FROM interviews i2
          WHERE i2.user_id = u.id AND i2.status = 'completed'
          LIMIT 5
        ) as recent_interviews
      FROM users u
      LEFT JOIN interviews i ON u.id = i.user_id
      WHERE u.role = 'USER'
    `;

    const params = [];
    let paramIndex = 1;

    // Add search filter
    if (search) {
      query += ` AND (u.name ILIKE $${paramIndex} OR u.email ILIKE $${paramIndex})`;
      params.push(`%${search}%`);
      paramIndex++;
    }

    // Add status filter
    if (status) {
      if (status === 'active') {
        query += ` AND EXISTS (SELECT 1 FROM interviews i2 WHERE i2.user_id = u.id AND i2.status IN ('in_progress', 'pending'))`;
      } else if (status === 'completed') {
        query += ` AND EXISTS (SELECT 1 FROM interviews i2 WHERE i2.user_id = u.id AND i2.status = 'completed')`;
      } else if (status === 'new') {
        query += ` AND NOT EXISTS (SELECT 1 FROM interviews i2 WHERE i2.user_id = u.id AND i2.status = 'completed')`;
      }
    }

    // Group by user
    query += ` GROUP BY u.id, u.name, u.email, u.created_at`;

    // Order by
    const validSortColumns = ['avg_score', 'total_interviews', 'completed_interviews', 'joined_date', 'name', 'last_interview_date'];
    const sortColumn = validSortColumns.includes(sortBy) ? sortBy : 'avg_score';
    const order = sortOrder.toUpperCase() === 'ASC' ? 'ASC' : 'DESC';
    query += ` ORDER BY ${sortColumn} ${order}, u.name ASC`;

    const result = await pool.query(query, params);

    // Process each candidate's data
    const candidates = result.rows.map(candidate => ({
      id: candidate.id,
      name: candidate.name,
      email: candidate.email,
      joinedDate: candidate.joined_date,
      totalInterviews: parseInt(candidate.total_interviews) || 0,
      completedInterviews: parseInt(candidate.completed_interviews) || 0,
      avgScore: Math.round(candidate.avg_score || 0),
      highestScore: Math.round(candidate.highest_score || 0),
      lowestScore: Math.round(candidate.lowest_score || 0),
      lastInterviewDate: candidate.last_interview_date,
      activeSessions: parseInt(candidate.active_sessions) || 0,
      recentInterviews: candidate.recent_interviews || [],
      status: candidate.completed_interviews > 0 ? 'completed' : 
              candidate.active_sessions > 0 ? 'active' : 'new'
    }));

    // Calculate ranking
    const rankedCandidates = candidates.map((c, index) => ({
      ...c,
      rank: index + 1,
      percentile: candidates.length > 0 
        ? Math.round(((candidates.length - index) / candidates.length) * 100) 
        : 0
    }));

    // Get summary stats
    const summary = {
      totalCandidates: candidates.length,
      activeCandidates: candidates.filter(c => c.activeSessions > 0).length,
      completedCandidates: candidates.filter(c => c.completedInterviews > 0).length,
      newCandidates: candidates.filter(c => c.status === 'new').length,
      averageScore: candidates.length > 0 
        ? Math.round(candidates.reduce((sum, c) => sum + c.avgScore, 0) / candidates.length) 
        : 0,
      topPerformer: candidates.length > 0 ? candidates[0] : null,
      totalInterviews: candidates.reduce((sum, c) => sum + c.totalInterviews, 0),
      completedInterviews: candidates.reduce((sum, c) => sum + c.completedInterviews, 0)
    };

    res.json({
      success: true,
      data: {
        candidates: rankedCandidates,
        summary,
        filters: {
          search: search || '',
          status: status || 'all'
        }
      }
    });

  } catch (error) {
    console.error('Error fetching candidates:', error);
    res.status(500).json({ 
      success: false, 
      error: 'Failed to fetch candidates',
      details: error.message 
    });
  }
});

// =============================================
// GET CANDIDATE DETAILS
// =============================================
router.get('/candidates/:id', auth, checkRole(['RECRUITER', 'ADMIN']), async (req, res) => {
  try {
    const { id } = req.params;

    // Get candidate details
    const userResult = await pool.query(
      `SELECT id, name, email, role, created_at 
       FROM users 
       WHERE id = $1 AND role = 'USER'`,
      [id]
    );

    if (userResult.rows.length === 0) {
      return res.status(404).json({ 
        success: false, 
        error: 'Candidate not found' 
      });
    }

    const candidate = userResult.rows[0];

    // Get all interviews
    const interviewsResult = await pool.query(
      `SELECT 
        id, interview_type, domain, difficulty, 
        score, status, feedback, 
        created_at, start_time, end_time,
        submission_type,
        communication_score,
        confidence_score,
        technical_score,
        professionalism_score
       FROM interviews 
       WHERE user_id = $1 
       ORDER BY created_at DESC`,
      [id]
    );

    // Process interviews
    const interviews = interviewsResult.rows.map(interview => {
      let feedback = interview.feedback;
      if (typeof feedback === 'string') {
        try { feedback = JSON.parse(feedback); } catch (e) { feedback = null; }
      }
      
      return {
        id: interview.id,
        type: interview.interview_type,
        domain: interview.domain,
        difficulty: interview.difficulty,
        score: interview.score,
        status: interview.status,
        feedback: feedback,
        submissionType: interview.submission_type || 'full',
        communicationScore: interview.communication_score,
        confidenceScore: interview.confidence_score,
        technicalScore: interview.technical_score,
        professionalismScore: interview.professionalism_score,
        createdAt: interview.created_at,
        startTime: interview.start_time,
        endTime: interview.end_time
      };
    });

    // Calculate stats
    const completed = interviews.filter(i => i.status === 'completed' && i.score);
    const scores = completed.map(i => i.score);
    
    const stats = {
      totalInterviews: interviews.length,
      completedInterviews: completed.length,
      avgScore: scores.length > 0 ? Math.round(scores.reduce((a, b) => a + b, 0) / scores.length) : 0,
      highestScore: scores.length > 0 ? Math.max(...scores) : 0,
      lowestScore: scores.length > 0 ? Math.min(...scores) : 0,
      activeSessions: interviews.filter(i => i.status === 'in_progress' || i.status === 'pending').length
    };

    // Get skill breakdown from feedback
    const skillBreakdown = {
      communication: { scores: [], count: 0 },
      confidence: { scores: [], count: 0 },
      technical: { scores: [], count: 0 },
      professionalism: { scores: [], count: 0 }
    };

    completed.forEach(i => {
      if (i.feedback) {
        if (i.feedback.communication_clarity !== undefined) {
          skillBreakdown.communication.scores.push(i.feedback.communication_clarity);
          skillBreakdown.communication.count++;
        }
        if (i.feedback.confidence !== undefined) {
          skillBreakdown.confidence.scores.push(i.feedback.confidence);
          skillBreakdown.confidence.count++;
        }
        if (i.feedback.technical_accuracy !== undefined) {
          skillBreakdown.technical.scores.push(i.feedback.technical_accuracy);
          skillBreakdown.technical.count++;
        }
        if (i.feedback.professionalism !== undefined || i.feedback.professionalism_score !== undefined) {
          const prof = i.feedback.professionalism || i.feedback.professionalism_score || 0;
          skillBreakdown.professionalism.scores.push(prof);
          skillBreakdown.professionalism.count++;
        }
      }
    });

    const skills = {};
    for (const [key, data] of Object.entries(skillBreakdown)) {
      skills[key] = {
        average: data.scores.length > 0 ? Math.round(data.scores.reduce((a, b) => a + b, 0) / data.scores.length) : 0,
        count: data.count,
        max: data.scores.length > 0 ? Math.max(...data.scores) : 0,
        min: data.scores.length > 0 ? Math.min(...data.scores) : 0
      };
    }

    res.json({
      success: true,
      data: {
        candidate: {
          id: candidate.id,
          name: candidate.name,
          email: candidate.email,
          joinedAt: candidate.created_at,
          role: candidate.role
        },
        stats,
        interviews,
        skills,
        summary: {
          totalScore: stats.avgScore,
          rating: stats.avgScore >= 85 ? 'Excellent' :
                  stats.avgScore >= 70 ? 'Good' :
                  stats.avgScore >= 55 ? 'Average' : 'Needs Improvement'
        }
      }
    });

  } catch (error) {
    console.error('Error fetching candidate details:', error);
    res.status(500).json({ 
      success: false, 
      error: 'Failed to fetch candidate details',
      details: error.message 
    });
  }
});

// =============================================
// SHORTLIST CANDIDATE
// =============================================
router.post('/candidates/:id/shortlist', auth, checkRole(['RECRUITER', 'ADMIN']), async (req, res) => {
  try {
    const { id } = req.params;
    const { notes = '' } = req.body;

    // Check if user exists and is a candidate
    const userResult = await pool.query(
      'SELECT id, name, email FROM users WHERE id = $1 AND role = $2',
      [id, 'USER']
    );

    if (userResult.rows.length === 0) {
      return res.status(404).json({ 
        success: false, 
        error: 'Candidate not found' 
      });
    }

    // Create or update shortlist entry
    const result = await pool.query(
      `INSERT INTO shortlist (candidate_id, recruiter_id, notes, created_at)
       VALUES ($1, $2, $3, CURRENT_TIMESTAMP)
       ON CONFLICT (candidate_id, recruiter_id) 
       DO UPDATE SET notes = EXCLUDED.notes, updated_at = CURRENT_TIMESTAMP
       RETURNING *`,
      [id, req.user.id, notes]
    );

    res.json({
      success: true,
      data: {
        message: 'Candidate shortlisted successfully',
        shortlist: result.rows[0]
      }
    });

  } catch (error) {
    console.error('Error shortlisting candidate:', error);
    res.status(500).json({ 
      success: false, 
      error: 'Failed to shortlist candidate',
      details: error.message 
    });
  }
});

// =============================================
// REMOVE FROM SHORTLIST
// =============================================
router.delete('/candidates/:id/shortlist', auth, checkRole(['RECRUITER', 'ADMIN']), async (req, res) => {
  try {
    const { id } = req.params;

    const result = await pool.query(
      'DELETE FROM shortlist WHERE candidate_id = $1 AND recruiter_id = $2 RETURNING *',
      [id, req.user.id]
    );

    if (result.rows.length === 0) {
      return res.status(404).json({ 
        success: false, 
        error: 'Candidate not in shortlist' 
      });
    }

    res.json({
      success: true,
      data: {
        message: 'Candidate removed from shortlist'
      }
    });

  } catch (error) {
    console.error('Error removing from shortlist:', error);
    res.status(500).json({ 
      success: false, 
      error: 'Failed to remove from shortlist',
      details: error.message 
    });
  }
});

// =============================================
// GET SHORTLISTED CANDIDATES
// =============================================
router.get('/shortlist', auth, checkRole(['RECRUITER', 'ADMIN']), async (req, res) => {
  try {
    const result = await pool.query(
      `SELECT 
        s.candidate_id,
        s.notes,
        s.created_at as shortlisted_at,
        u.name,
        u.email,
        (
          SELECT AVG(score) 
          FROM interviews 
          WHERE user_id = u.id AND status = 'completed'
        ) as avg_score,
        (
          SELECT COUNT(*) 
          FROM interviews 
          WHERE user_id = u.id AND status = 'completed'
        ) as completed_interviews
       FROM shortlist s
       JOIN users u ON s.candidate_id = u.id
       WHERE s.recruiter_id = $1
       ORDER BY s.created_at DESC`,
      [req.user.id]
    );

    const shortlist = result.rows.map(item => ({
      candidateId: item.candidate_id,
      name: item.name,
      email: item.email,
      avgScore: Math.round(item.avg_score || 0),
      completedInterviews: parseInt(item.completed_interviews) || 0,
      notes: item.notes,
      shortlistedAt: item.shortlisted_at
    }));

    res.json({
      success: true,
      data: {
        shortlist,
        total: shortlist.length
      }
    });

  } catch (error) {
    console.error('Error fetching shortlist:', error);
    res.status(500).json({ 
      success: false, 
      error: 'Failed to fetch shortlist',
      details: error.message 
    });
  }
});

module.exports = router;